File: backend/src/llm/providers/openai-provider.ts
import OpenAI from 'openai';
import { ModelProvider, ModelConfig, ModelResponse, StreamChunk } from '../model-provider';
import tiktoken from 'tiktoken';

export class OpenAIProvider extends ModelProvider {
  private client: OpenAI;
  private tokenizer: any;
  
  constructor(config: ModelConfig) {
    super(config);
  }
  
  async initialize(): Promise<void> {
    this.client = new OpenAI({
      apiKey: this.config.apiKey || process.env.OPENAI_API_KEY
    });
    
    // 토크나이저 초기화
    try {
      this.tokenizer = tiktoken.encoding_for_model(this.config.name as any);
    } catch {
      this.tokenizer = tiktoken.encoding_for_model('gpt-4');
    }
  }
  
  async generate(
    prompt: string,
    options?: Partial<ModelConfig>
  ): Promise<ModelResponse> {
    this.validateInput(prompt);
    const config = this.mergeConfig(options);
    
    const startTime = Date.now();
    
    try {
      const response = await this.client.chat.completions.create({
        model: config.name,
        messages: [{ role: 'user', content: prompt }],
        max_tokens: config.maxTokens,
        temperature: config.temperature,
        top_p: config.topP,
        frequency_penalty: config.frequencyPenalty,
        presence_penalty: config.presencePenalty,
        stop: config.stopSequences.length > 0 ? config.stopSequences : undefined
      });
      
      const choice = response.choices[0];
      const tokensUsed = response.usage?.total_tokens || 0;
      
      return {
        text: choice.message.content || '',
        tokensUsed,
        finishReason: choice.finish_reason || 'stop',
        metadata: {
          model: response.model,
          created: response.created,
          latency: Date.now() - startTime,
          promptTokens: response.usage?.prompt_tokens,
          completionTokens: response.usage?.completion_tokens
        },
        cost: this.getCostEstimate(
          response.usage?.prompt_tokens || 0,
          response.usage?.completion_tokens || 0
        )
      };
      
    } catch (error: any) {
      throw new Error(`OpenAI API error: ${error.message}`);
    }
  }
  
  async *streamGenerate(
    prompt: string,
    options?: Partial<ModelConfig>
  ): AsyncIterableIterator<StreamChunk> {
    this.validateInput(prompt);
    const config = this.mergeConfig(options);
    
    try {
      const stream = await this.client.chat.completions.create({
        model: config.name,
        messages: [{ role: 'user', content: prompt }],
        max_tokens: config.maxTokens,
        temperature: config.temperature,
        top_p: config.topP,
        frequency_penalty: config.frequencyPenalty,
        presence_penalty: config.presencePenalty,
        stop: config.stopSequences.length > 0 ? config.stopSequences : undefined,
        stream: true
      });
      
      for await (const chunk of stream) {
        const choice = chunk.choices[0];
        const content = choice?.delta?.content || '';
        
        yield {
          text: content,
          isComplete: choice?.finish_reason !== null,
          metadata: {
            finishReason: choice?.finish_reason
          }
        };
        
        if (choice?.finish_reason) {
          break;
        }
      }
      
    } catch (error: any) {
      throw new Error(`OpenAI streaming error: ${error.message}`);
    }
  }
  
  async embed(texts: string[]): Promise<number[][]> {
    try {
      const response = await this.client.embeddings.create({
        model: 'text-embedding-ada-002',
        input: texts
      });
      
      return response.data.map(item => item.embedding);
      
    } catch (error: any) {
      throw new Error(`OpenAI embedding error: ${error.message}`);
    }
  }
  
  estimateTokens(text: string): number {
    if (!this.tokenizer) {
      // 대략적인 추정 (1 토큰 ≈ 4 문자)
      return Math.ceil(text.length / 4);
    }
    
    return this.tokenizer.encode(text).length;
  }
  
  getCostEstimate(inputTokens: number, outputTokens: number): number {
    // OpenAI 가격표 (2024년 기준, 실제로는 동적으로 가져와야 함)
    const pricing: Record<string, { input: number; output: number }> = {
      'gpt-4': { input: 0.00003, output: 0.00006 },
      'gpt-4-turbo': { input: 0.00001, output: 0.00003 },
      'gpt-3.5-turbo': { input: 0.0000015, output: 0.000002 }
    };
    
    const modelPricing = pricing[this.config.name] || pricing['gpt-4'];
    
    return (inputTokens * modelPricing.input) + (outputTokens * modelPricing.output);
  }
}

File: backend/src/llm/providers/openai_provider.py
import openai
import os
from typing import AsyncIterator, Dict, Any, Optional, List
from ..base import ModelProvider, ModelResponse


class OpenAIProvider(ModelProvider):
    """OpenAI 모델 프로바이더"""

    async def initialize(self) -> None:
        self.client = openai.AsyncOpenAI(api_key=os.getenv("OPENAI_API_KEY"))

    async def generate(
        self, prompt: str, options: Optional[Dict[str, Any]] = None
    ) -> ModelResponse:
        options = options or {}

        response = await self.client.chat.completions.create(
            model=self.config.name,
            messages=[{"role": "user", "content": prompt}],
            max_tokens=options.get("max_tokens", self.config.max_tokens),
            temperature=options.get("temperature", self.config.temperature),
            top_p=options.get("top_p", self.config.top_p),
            frequency_penalty=options.get(
                "frequency_penalty", self.config.frequency_penalty
            ),
            presence_penalty=options.get(
                "presence_penalty", self.config.presence_penalty
            ),
            stop=options.get("stop", self.config.stop_sequences),
        )

        choice = response.choices[0]
        return ModelResponse(
            text=choice.message.content,
            tokens_used=response.usage.total_tokens,
            finish_reason=choice.finish_reason,
            metadata={"model": response.model, "created": response.created},
        )

    async def stream_generate(
        self, prompt: str, options: Optional[Dict[str, Any]] = None
    ) -> AsyncIterator[str]:
        options = options or {}

        stream = await self.client.chat.completions.create(
            model=self.config.name,
            messages=[{"role": "user", "content": prompt}],
            stream=True,
            **options
        )

        async for chunk in stream:
            if chunk.choices[0].delta.content:
                yield chunk.choices[0].delta.content

    async def embed(self, texts: List[str]) -> List[List[float]]:
        response = await self.client.embeddings.create(
            model="text-embedding-ada-002", input=texts
        )
        return [data.embedding for data in response.data]

    def estimate_tokens(self, text: str) -> int:
        return len(text) // 4  # 대략적 추정

    def get_cost_estimate(self, input_tokens: int, output_tokens: int) -> float:
        # GPT-4 기준 비용
        return (input_tokens * 0.00003) + (output_tokens * 0.00006)
